rkt/image/namefetcher.go

```diff
@@ -0,0 +1,257 @@
+	// for this we need to clone "a" and call
+	// maybeOverrideAscFetcherWithRemote on the clone
+	aciURL := ep.ACIEndpoints[0].ACI
+	ascURL := ep.ACIEndpoints[0].ASC
+	stderr("remote fetching from URL %q", aciURL)
+	f.maybeOverrideAscFetcherWithRemote(ascURL, a)
+	return f.fetchImageFromSingleEndpoint(app, aciURL, a, latest)
+}
+
+func (f *nameFetcher) fetchImageFromSingleEndpoint(app *discovery.App, aciURL string, a *asc, latest bool) (string, error) {
+	if f.Debug {
+		stderr("fetching image from %s", aciURL)
+	}
+
+	aciFile, cd, err := f.fetch(app, aciURL, a)
+	if err != nil {
+		return "", err
+	}
+	defer aciFile.Close()
+
+	key, err := f.S.WriteACI(aciFile, latest)
+	if err != nil {
+		return "", err
+	}
+
+	rem := store.NewRemote(aciURL, a.Location)
+	rem.BlobKey = key
+	rem.DownloadTime = time.Now()
+	rem.ETag = cd.ETag
+	rem.CacheMaxAge = cd.MaxAge
+	err = f.S.WriteRemote(rem)
+	if err != nil {
+		return "", err
+	}
+
+	return key, nil
+}
+
+func (f *nameFetcher) fetch(app *discovery.App, aciURL string, a *asc) (readSeekCloser, *cacheData, error) {
+	if f.InsecureFlags.SkipTlsCheck() && f.Ks != nil {
+		stderr("warning: TLS verification has been disabled")
+	}
+	if f.InsecureFlags.SkipImageCheck() && f.Ks != nil {
+		stderr("warning: image signature verification has been disabled")
+	}
+
+	u, err := url.Parse(aciURL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing ACI url: %v", err)
+	}
+
+	if f.InsecureFlags.SkipImageCheck() || f.Ks == nil {
+		o := f.getHttpOps()
+		aciFile, cd, err := o.DownloadImage(u)
+		if err != nil {
+			return nil, nil, err
+		}
+		return aciFile, cd, nil
+	}
+
+	return f.fetchVerifiedURL(app, u, a)
+}
+
+func (f *nameFetcher) fetchVerifiedURL(app *discovery.App, u *url.URL, a *asc) (readSeekCloser, *cacheData, error) {
+	appName := app.Name.String()
+	f.maybeFetchPubKeys(appName)
+
+	o := f.getHttpOps()
+	ascFile, retry, err := o.DownloadSignature(a)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() { maybeClose(ascFile) }()
+
+	if !retry {
+		if err := f.checkIdentity(appName, ascFile); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	aciFile, cd, err := o.DownloadImage(u)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() { maybeClose(aciFile) }()
+
+	if retry {
+		ascFile, err = o.DownloadSignatureAgain(a)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if err := f.validate(appName, aciFile, ascFile); err != nil {
+		return nil, nil, err
+	}
+	retAciFile := aciFile
+	aciFile = nil
+	return retAciFile, cd, nil
+}
+
+func (f *nameFetcher) maybeFetchPubKeys(appName string) {
+	if f.TrustKeysFromHttps && !f.InsecureFlags.SkipTlsCheck() {
+		m := &pubkey.Manager{
+			AuthPerHost:        f.Headers,
+			InsecureAllowHttp:  false,
+			TrustKeysFromHttps: true,
+			Ks:                 f.Ks,
+			Debug:              f.Debug,
+		}
+		pkls, err := m.GetPubKeyLocations(appName)
+		// TODO(krnowak): Should errors below be fatal?
```

Comment:
I don't think so, if fetching the public keys fails but we already trust the key, we should be able to run the image anyway.